### PR TITLE
CustomBgpIpConfigurations in gateway connections

### PR DESCRIPTION
**Pull request by Khushboo-Baheti**
## Description
Adding New Cmdlets to create PSGatewayCustomBgpIpAddressIpConfiguration.
Adding new parameter GatewayCustomBgpIpAddre to Set-AzVirtualNetworkGatewayConnection and New-AzVirtualNetworkGatewayConnection.
Adding new parameter VpnGatewayCustomBgpIpAddress to New-AzVpnSiteLinkConnection.

https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1306
## Checklist

- [X] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Network/Network/help/New-AzGatewayCustomBgpIpConfigurationObject.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Network.dll-Help.xml
+Module Name: Az.Network
+online version: https://docs.microsoft.com/en-us/powershell/module/az.network/new-azgatewaycustombgpipconfigurationobject
```
Comment (VeryEarly) [inline src/Network/Network/help/New-AzGatewayCustomBgpIpConfigurationObject.md:4]:
```suggestion
online version: https://learn.microsoft.com/powershell/module/az.network/new-azgatewaycustombgpipconfigurationobject
```

File: src/Network/Network/ChangeLog.md [hunk 2]
```diff
@@ -24,6 +24,8 @@
 * Fixed a bug that does not enable to set Perform SNAT to Always
 * Fixed the incorrect type of `-TotalBytesPerSession` in `New-AzNetworkWatcherPacketCapture`
 * Added support for connection flushing in network security group which when enabled, re-evaluates flows when rules are updated
+* Added New-AzGatewayCustomBgpIpConfigurationObject command
+* Updated  New-AzVirtualNetworkGatewayConnection, Set-AzVirtualNetworkGatewayConnection and New-AzVpnSiteLinkConnection to support GatewayCustomBgpIpConfiguration.
     - `New-AzNetworkSecurityGroup`
```
Comment (VeryEarly) [inline src/Network/Network/ChangeLog.md:29]: please insert your changelog under this line

File: src/Network/Network/ChangeLog.md [hunk 3]
```diff
@@ -24,6 +24,8 @@
 * Fixed a bug that does not enable to set Perform SNAT to Always
 * Fixed the incorrect type of `-TotalBytesPerSession` in `New-AzNetworkWatcherPacketCapture`
 * Added support for connection flushing in network security group which when enabled, re-evaluates flows when rules are updated
+* Added New-AzGatewayCustomBgpIpConfigurationObject command
```
Comment (VeryEarly) [inline src/Network/Network/ChangeLog.md:27]:
```suggestion
* Added `New-AzGatewayCustomBgpIpConfigurationObject` command
```

File: src/Network/Network/ChangeLog.md [hunk 4]
```diff
@@ -24,6 +24,8 @@
...
+* Updated  New-AzVirtualNetworkGatewayConnection, Set-AzVirtualNetworkGatewayConnection and New-AzVpnSiteLinkConnection to support GatewayCustomBgpIpConfiguration.
```
Comment (VeryEarly) [inline src/Network/Network/ChangeLog.md:28]:
```suggestion
* Updated  `New-AzVirtualNetworkGatewayConnection`, `Set-AzVirtualNetworkGatewayConnection` and `New-AzVpnSiteLinkConnection` to support GatewayCustomBgpIpConfiguration.
```

File: src/Network/Network/VirtualNetworkGatewayConnection/GetAzureVirtualNetworkGatewayConnectionCommand.cs [hunk 5]
```diff
@@ -57,7 +57,7 @@ public override void Execute()
                 var connectionPage = this.VirtualNetworkGatewayConnectionClient.List(this.ResourceGroupName);
 
                 // Get all resources by polling on next page link
-                var connectionList = ListNextLink<VirtualNetworkGatewayConnection>.GetAllResourcesByPollingNextLink(connectionPage, this.VirtualNetworkGatewayConnectionClient.ListNext);
+                var connectionList = ListNextLink<Management.Network.Models.VirtualNetworkGatewayConnection>.GetAllResourcesByPollingNextLink(connectionPage, this.VirtualNetworkGatewayConnectionClient.ListNext);
```
Comment (VeryEarly) [inline src/Network/Network/VirtualNetworkGatewayConnection/GetAzureVirtualNetworkGatewayConnectionCommand.cs:60]: is this change added by your IDE? please remove if not necessary

File: src/Network/Network/VirtualNetworkGatewayConnection/NewAzureVirtualNetworkGatewayConnectionCommand.cs [hunk 6]
```diff
@@ -300,6 +307,16 @@ private PSVirtualNetworkGatewayConnection CreateVirtualNetworkGatewayConnection(
                 }
             }
 
+            if (this.GatewayCustomBgpIpAddress != null)
+            {
+                vnetGatewayConnection.GatewayCustomBgpIpAddresses = new List<PSGatewayCustomBgpIpConfiguration>();
+
+                foreach (var reqaddress in this.GatewayCustomBgpIpAddress)
```
Comment (VeryEarly) [inline src/Network/Network/VirtualNetworkGatewayConnection/NewAzureVirtualNetworkGatewayConnectionCommand.cs:314]: NIT:
vnetGatewayConnection.GatewayCustomBgpIpAddresses.AddRange(GatewayCustomBgpIpAddress);

File: src/Network/Network/VirtualNetworkGatewayConnection/VirtualNetworkGatewayConnectionBaseCmdlet.cs [hunk 7]
```diff
@@ -61,7 +61,7 @@ public PSVirtualNetworkGatewayConnection GetVirtualNetworkGatewayConnection(stri
             return psVirtualNetworkGatewayConnection;
         }
 
-        public PSVirtualNetworkGatewayConnection ToPsVirtualNetworkGatewayConnection(VirtualNetworkGatewayConnection vnetGatewayConnection)
+        public PSVirtualNetworkGatewayConnection ToPsVirtualNetworkGatewayConnection(Management.Network.Models.VirtualNetworkGatewayConnection vnetGatewayConnection)
```
Comment (VeryEarly) [inline src/Network/Network/VirtualNetworkGatewayConnection/VirtualNetworkGatewayConnectionBaseCmdlet.cs:64]: same as GetAzureVirtualNetworkGatewayConnectionCommand.cs

**Review by VeryEarly — CHANGES_REQUESTED** [review]
please see comments

File: src/Network/Network/Cortex/VpnConnection/NewAzVpnSiteLinkConnectionCommand.cs [hunk 8]
```diff
@@ -85,6 +86,13 @@ public class NewAzVpnSiteLinkConnectionCommand : NetworkBaseCmdlet
             HelpMessage = "The list of egress  NAT rules that are associated with this link Connection.")]
         public PSResourceId[] EgressNatRule { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The GatewayCustomBgpIpAddress of Vpngateway used in this link connection.")]
+        [ValidateNotNullOrEmpty]
+        public PSGatewayCustomBgpIpConfiguration[] VpnGatewayCustomBgpAddresses { get; set; }
```
Comment (VeryEarly) [inline src/Network/Network/Cortex/VpnConnection/NewAzVpnSiteLinkConnectionCommand.cs:94]:
```suggestion
        public PSGatewayCustomBgpIpConfiguration[] VpnGatewayCustomBgpAddress { get; set; }
```

File: src/Network/Network/Models/Cortex/PSVpnSiteLinkConnection.cs [hunk 9]
```diff
@@ -49,6 +49,9 @@ public class PSVpnSiteLinkConnection : PSChildResource
         [Ps1Xml(Label = "EgressNatRules", Target = ViewControl.Table)]
         public List<PSResourceId> EgressNatRules { get; set; }
 
+        [Ps1Xml(Label = "VpnGatewayCustomBgpAddresses", Target = ViewControl.Table)]
+        public List<PSGatewayCustomBgpIpConfiguration> VpnGatewayCustomBgpAddresses { get; set; }
```
Comment (VeryEarly) [inline src/Network/Network/Models/Cortex/PSVpnSiteLinkConnection.cs:53]:
```suggestion
        public List<PSGatewayCustomBgpIpConfiguration> VpnGatewayCustomBgpAddress { get; set; }
```

File: src/Network/Network/Models/PSVirtualNetworkGatewayConnection.cs [hunk 10]
```diff
@@ -83,6 +83,9 @@ public class PSVirtualNetworkGatewayConnection : PSTopLevelResource
         [Ps1Xml(Label = "EgressNatRules", Target = ViewControl.Table)]
         public List<PSResourceId> EgressNatRules { get; set; }
 
+        [Ps1Xml(Label = "GatewayCustomBgpIpAddresses", Target = ViewControl.Table)]
+        public List<PSGatewayCustomBgpIpConfiguration> GatewayCustomBgpIpAddresses { get; set; }
```
Comment (VeryEarly) [inline src/Network/Network/Models/PSVirtualNetworkGatewayConnection.cs:87]:
```suggestion
        public List<PSGatewayCustomBgpIpConfiguration> GatewayCustomBgpIpAddress { get; set; }
```

**Comment by VeryEarly** [comment]
please find analyze-netcore3.1 -> staticAnalysisResults -> ExampleIssues.csv in
https://dev.azure.com/azclitools/public/_build/results?buildId=38863&view=artifacts&pathAsName=false&type=publishedArtifacts
and fix examples.

**Comment by Khushboo-Baheti** [comment]
> please find analyze-netcore3.1 -> staticAnalysisResults -> ExampleIssues.csv in https://dev.azure.com/azclitools/public/_build/results?buildId=38863&view=artifacts&pathAsName=false&type=publishedArtifacts and fix examples.

Not able to find anything here

**Review by VeryEarly — APPROVED** [review]

**Pull request merged**